quinn-proto/src/connection/pacing.rs

```diff
@@ -0,0 +1,125 @@
+//! Pacing of packet transmissions.
+
+use std::time::{Duration, Instant};
+
+use tracing::{debug, warn};
+
+/// A simple token-bucket pacer. The bucket starts full and has an adjustable capacity. Once the
+/// bucket is empty, further transmission is blocked. The bucket refills at a rate slightly faster
+/// than one contestion window per RTT.
+pub struct Pacer {
+    capacity: u64,
+    tokens: u64,
+    prev: Instant,
+}
+
+impl Pacer {
+    /// Obtains a new [`Pacer`].
+    pub fn new(capacity: u64, now: Instant) -> Self {
+        Self {
+            capacity,
+            tokens: capacity,
+            prev: now,
+        }
+    }
+
+    /// Record that a packet has been transmitted.
+    pub fn on_transmit(&mut self, packet_length: u16) {
+        self.tokens = self.tokens.saturating_sub(packet_length.into())
+    }
+
+    /// Return how long we need to wait before sending a packet.
+    ///
+    /// If we can send a packet right away, this returns `None`. Otherwise, returns `Some(d)`,
+    /// where `d` is the time before this function should be called again.
+    ///
+    /// The 5/4 ratio used here comes from the suggestion that N = 1.25 in the draft IETF RFC for
+    /// QUIC.
+    pub fn delay(
+        &mut self,
+        smoothed_rtt: Duration,
+        mtu: u16,
+        window: u64,
```

Comment:
It doesn't look like `window`'s natural type for this algorithm is `u64` -- it gets compared to `u32::max_value()`, cast to `f64` for the token calculation and then case to `u32` again in the return value. Similarly, it looks like `mtu` gets cast to `u64` at all usage sites. Would be better to cast these once early on in the method or take their argument in the "right" type directly, depending on whether the caller or this method should handle any conversion edge cases.

Reply:
Most of this method *is* conversion edge cases :smile:.  Another option is to just cast to `f64` at the start of the method and then cast back at the end, which would simplify the code significantly.

Reply:
That'd work for me, too.

Reply:
I fixed it by doing the calculation up-front.  Using floating point everywhere made me a bit nervous about loss of precision.